src/model/encoding/convertFromHTMLToContentBlocks.js

```diff
@@ -564,60 +607,128 @@ function getChunkForHTML(
   }
 
   return {chunk, entityMap: newEntityMap};
-}
+};
 
-function convertFromHTMLtoContentBlocks(
+const convertChunkToContentBlocks = (
+  chunk: Chunk,
+  experimentalTreeDataSupport: boolean,
+): ?Array<BlockNodeRecord> => {
+  if (!chunk || !chunk.text || !Array.isArray(chunk.blocks)) {
+    return null;
+  }
+
+  const initialState = {
+    cacheRef: {},
+    contentBlocks: [],
+  };
+
+  let start = 0;
+
+  const {blocks: rawBlocks, inlines: rawInlines, entities: rawEntities} = chunk;
+
+  const BlockNodeRecord = experimentalTreeDataSupport
+    ? ContentBlockNode
+    : ContentBlock;
+
+  return chunk.text.split('\r').reduce((acc, textBlock, index) => {
+    // Make absolutely certain that our text is acceptable.
+    textBlock = sanitizeDraftText(textBlock);
+
+    const block = rawBlocks[index];
+    const end = start + textBlock.length;
+    const inlines = rawInlines.slice(start, end);
+    const entities = rawEntities.slice(start, end);
+    const characterList = List(
+      inlines.map((style, index) => {
+        const data = {style, entity: (null: ?string)};
+        if (entities[index]) {
+          data.entity = entities[index];
+        }
+        return CharacterMetadata.create(data);
+      }),
+    );
+    start = end + 1;
+
+    const {depth, type, parent} = block;
+
+    const key = block.key || generateRandomKey();
+
+    const blockNode = new BlockNodeRecord({
+      key,
+      parent,
+      type,
+      depth,
+      text: textBlock,
+      characterList,
+      prevSibling:
+        index === 0 || rawBlocks[index - 1].parent !== parent
+          ? null
+          : rawBlocks[index - 1].key,
+      nextSibling:
+        index === rawBlocks.length - 1 || rawBlocks[index + 1].parent !== parent
+          ? null
+          : rawBlocks[index + 1].key,
+    });
+
+    // childrens add themselves to their parents since we are iterating in order
+    if (parent) {
+      const parentIndex = acc.cacheRef[parent];
+      const parentRecord = acc.contentBlocks[parentIndex];
+      acc.contentBlocks[parentIndex] = parentRecord.set(
+        'children',
+        parentRecord.children.push(key),
+      );
+    }
+
+    // insert node
+    acc.contentBlocks.push(blockNode);
+
+    // cache ref for building links
+    acc.cacheRef[blockNode.key] = index;
+
+    return acc;
+  }, initialState).contentBlocks;
+};
+
+/**
+ * Note for `experimentalTreeDataSupport`:
+ *
+ * This is unstable and not part of the public API and should not be used by
+ * production systems. This functionality may be update/removed without notice.
+ */
+const convertFromHTMLtoContentBlocks = (
   html: string,
   DOMBuilder: Function = getSafeBodyFromHTML,
   blockRenderMap?: DraftBlockRenderMap = DefaultDraftBlockRenderMap,
-): ?{contentBlocks: ?Array<BlockNodeRecord>, entityMap: EntityMap} {
+  experimentalTreeDataSupport?: boolean = false,
```

Comment:
The gk task is the next task i want to cover on a PR this was to avoid bloating this already complicated PR to review.

My goal is having a follow up PR which will move all the `experimentalTreeDataSupport` flags to be controlled by a `GK` flag, was hoping to discuss a bit in person with you prior to starting it